scuttlebutt-handshake/src/main/java/net/consensys/cava/scuttlebutt/handshake/SecureScuttlebuttStream.java

```diff
@@ -100,29 +100,38 @@ private Bytes decrypt(Bytes message, SecretBox.Key key, MutableBytes nonce, bool
   }
 
   private Bytes decryptMessage(Bytes message, SecretBox.Key key, MutableBytes nonce) {
-    if (message.size() < 34) {
-      return null;
-    }
     MutableBytes snapshotNonce = nonce.mutableCopy();
-    SecretBox.Nonce headerNonce = SecretBox.Nonce.fromBytes(snapshotNonce);
-    SecretBox.Nonce bodyNonce = SecretBox.Nonce.fromBytes(snapshotNonce.increment());
-    Bytes decryptedHeader = SecretBox.decrypt(message.slice(0, 34), key, headerNonce);
 
-    if (decryptedHeader == null) {
-      throw new StreamException("Failed to decrypt message header");
-    }
+    try {
+      try (SecretBox.Nonce headerNonce = SecretBox.Nonce.fromBytes(snapshotNonce);
+          SecretBox.Nonce bodyNonce = SecretBox.Nonce.fromBytes(snapshotNonce.increment());) {
+        if (message.size() < 34) {
+          return null;
+        }
 
-    int bodySize = ((decryptedHeader.get(0) & 0xFF) << 8) + (decryptedHeader.get(1) & 0xFF);
-    if (message.size() < bodySize + 34) {
-      return null;
-    }
-    Bytes body = message.slice(34, bodySize);
-    Bytes decryptedBody = SecretBox.decrypt(Bytes.concatenate(decryptedHeader.slice(2), body), key, bodyNonce);
-    if (decryptedBody == null) {
-      throw new StreamException("Failed to decrypt message");
+        Bytes decryptedHeader = SecretBox.decrypt(message.slice(0, 34), key, headerNonce);
+
+        if (decryptedHeader == null) {
+          throw new StreamException("Failed to decrypt message header");
+        }
+
+        int bodySize = ((decryptedHeader.get(0) & 0xFF) << 8) + (decryptedHeader.get(1) & 0xFF);
+        if (message.size() < bodySize + 34) {
+          return null;
+        }
+        Bytes body = message.slice(34, bodySize);
+        Bytes decryptedBody = SecretBox.decrypt(Bytes.concatenate(decryptedHeader.slice(2), body), key, bodyNonce);
+        if (decryptedBody == null) {
+          throw new StreamException("Failed to decrypt message");
+        }
+        nonce.increment().increment();
+
+        return decryptedBody;
+
+      }
+    } catch (Exception e) {
+      throw new StreamException(e.getMessage());
```

Comment:
Can we be more specific - Exception is very broad.
Also can we throw StreamException(e) so we get the whole stacktrace?